[swift5] Cancellable requests

DavidHorvath opened the pull request:
resolve #8658

Previously the generated clients didn't support **cancellation of requests**

## I've added request cancellation for:
- Both URLSession & Alamofire library
- Combine configuration
- RxSwift configuration
- Result configuration
- AsyncAwait configuration

For these clients I have updated the RequestBuilder's execute function so it returns an optional `URLSessionDataTask`.
I've also added `@discardableResult` to keep backward compatibility.

### Default
I've added to return the `URLSessionDataTask` so the user will be able to control the request's lifecycle.
``` swift
    @discardableResult
    open class func addPet(body: Pet, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue, completion: @escaping ((_ data: Void?, _ error: Error?) -> Void)) -> URLSessionDataTask? {
        return addPetWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
            switch result {
            case .success:
                completion((), nil)
            case let .failure(error):
                completion(nil, error)
            }
        }
    }
```

### Result
I've added to return the `URLSessionDataTask` so the user will be able to control the request's lifecycle.
``` swift
    @discardableResult
    open class func addPet(body: Pet, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue, completion: @escaping ((_ result: Swift.Result<Void, ErrorResponse>) -> Void)) -> URLSessionTask? {
        return addPetWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
            switch result {
            case .success:
                completion(.success(()))
            case let .failure(error):
                completion(.failure(error))
            }
        }
    }
```

### Combine
I've implemented [handeEvents(receiveCompletion:](https://developer.apple.com/documentation/combine/fail/handleevents(receivesubscription:receiveoutput:receivecompletion:receivecancel:receiverequest:)) where I cancel the returned `URLSessionDataTask` when the stream is cancelled
``` swift
open class func addPet(body: Pet, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) -> AnyPublisher<Void, Error> {
        var dataTask: URLSessionDataTask?
        return Future<Void, Error> { promise in
            dataTask = addPetWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
                switch result {
                case .success:
                    promise(.success(()))
                case let .failure(error):
                    promise(.failure(error))
                }
            }
        }
        .handleEvents(receiveCancel: {
            dataTask?.cancel()
        })
        .eraseToAnyPublisher()
    }
```

### RxSwift
I've added to cancel the returned `URLSessionDataTask` in `Disposables.create` when the stream is cancelled
``` swift
    open class func addPet(body: Pet, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) -> Observable<Void> {
        return Observable.create { observer -> Disposable in
            let dataTask = addPetWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
                switch result {
                case .success:
                    observer.onNext(())
                case let .failure(error):
                    observer.onError(error)
                }
                observer.onCompleted()
            }
            
            return Disposables.create {
                dataTask?.cancel()
            }
        }
    }
```

### Async Await
I've added to cancel the returned `URLSessionDataTask` in `withTaskCancellationHandler.onCancel:` 
``` swift
    open class func addPet(body: Pet, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) async throws {
        var task: URLSessionTask?
        return try await withTaskCancellationHandler {
            try Task.checkCancellation()
            return try await withCheckedThrowingContinuation { continuation in
                guard !Task.isCancelled else {
                  continuation.resume(throwing: CancellationError())
                  return
                }

                task = addPetWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
                    switch result {
                    case .success:
                        continuation.resume(returning: ())
                    case let .failure(error):
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { [task] in
            task?.cancel()
        }
    }
```



### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request. @4brunu

DavidHorvath commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -157,8 +161,9 @@ extension {{projectName}}API {
     @available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
     {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} class func {{operationId}}({{#allParams}}{{paramName}}: {{#isEnum}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{^isContainer}}{{{datatypeWithEnum}}}_{{operationId}}{{/isContainer}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{/isEnum}}{{^required}}? = nil{{/required}}{{^-last}}, {{/-last}}{{/allParams}}{{#hasParams}}, ...
+        {{#useURLSession}}var dataTask: URLSessionDataTask?{{/useURLSession}}
         return Future<{{{returnType}}}{{^returnType}}Void{{/returnType}}, Error> { promise in
```
@4brunu I have one question about this.
Don't we want to wrap this `Future publisher` into a [Deferred publisher](https://developer.apple.com/documentation/combine/deferred)?
So it would work like Rx's `Single` by having it defer its execution until it receives a subscriber.

4brunu replied:
I don't have much experience with `RxSwift` or `Combine`.
Is the current `RxSwift` integration lazy (defer its execution until it receives a subscriber)?

DavidHorvath replied:
@4brunu Yes. RxSwift Observable's only call when you subscribe to the stream.
In Combine publishers are working the same way except Future. We can only achieve the same if we would wrap Future in Deferred.

So currently if we just write this:
``` swift
let newPet = Pet(id:<id>, category: <category>, name: <name>, photoUrls: [])

let addPetPublisher = PetAPI.addPet(body: newPet)
```
The request will be called without subscribing to the stream with `.sink`

4brunu replied:
This is the way it is, because when I introduced support for `Combine` I didn't have much experience with it, and Future looked like a good fit.
I think we can also make `Combine` lazy, because that would make the API more flexible, by only starting the request when the user want's.
What do you think?

DavidHorvath replied:
Yes, I think we should do that 👍

4brunu replied:
Can this behaviour change create any bugs for the users that already implement Combine in their projects?
For example a network request that never happens because of this change?

DavidHorvath replied:
Unfortunately, yes 😕

DavidHorvath replied:
If somebody used it without subscribing to the stream, it won't be called

4brunu replied:
Hum, let's keep that for a different PR then, and keep this one with just the cancellation, to avoid any friction to merge it.

4brunu commented:
Hey @DavidHorvath, this is another great PR, thanks 👍 
Here are some thoughts.

Instead of returning an `URLSessionDataTask` I think we should return a `URLSessionTask` because it's less specialised.

We should also support cancellation in Alamofire, by replace the following code 
```        
if let encoding = encoding {
    let request = makeRequest(manager: manager, method: xMethod, encoding: encoding, headers: headers)
    if let onProgressReady = self.onProgressReady {
        onProgressReady(request.uploadProgress)
    }
    processRequest(request: request, managerId, apiResponseQueue, completion)
    return request.task
}

return nil
```
In the following class
https://github.com/OpenAPITools/openapi-generator/blob/f596b32316aea8486ceaa822647185621fae652a/modules/openapi-generator/src/main/resources/swift5/libraries/alamofire/AlamofireImplementations.mustache#L138-L144

Those two changes should a lot of conditions in the templates.

We should also support request cancelation in `result` by returning the `URLSessionTask` and `async/await` by using the `Task` but we need to explore this a bit more, because I don't have much experience with `async/await`.
https://www.hackingwithswift.com/quick-start/concurrency/how-to-cancel-a-task
https://desiatov.com/swift-structured-concurrency-introduction/
https://betterprogramming.pub/swiftui-concurrency-understand-cooperative-task-cancellation-b71cc69483e0

4brunu commented on modules/openapi-generator/src/main/resources/swift5/APIs.mustache:
```diff
@@ -56,8 +56,11 @@ import Vapor
     }
-
-    {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} func execute(_ apiResponseQueue: DispatchQueue = {{projectName}}API.apiResponseQueue, _ completion: @escaping (_ result: Swift.Result<Response<T>, ErrorResponse>) -> Void) { }
+    {{#useURLSession}}{{^useVapor}}{{^usePromiseKit}}{{^useResult}}{{^useAsyncAwait}}
```
@DavidHorvath do we need those conditions?
I don't we should return the `URLSessionDataTask` in every configuration, and then in the `api.mustache`we add the conditions, case by case.

DavidHorvath replied:
I've only added this condition because I only knew how I need to cancel request for URLSession with Default, Combine or RxSwift configurations and I didn't want to update the others 🙂 

But with your comments and links above, I have a direction to do it.
So if I update the other configurations as well, we won't need this

4brunu replied:
Sure, and you did great, because honestly in the past I looked at this big task and felt a bit discouraged.
If you need any help please let me know 👍

4brunu commented:
Looks like the way to cancel an async/await function is https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md#cancellation-handlers

Here is the example in the like above.

```
func download(url: URL) async throws -> Data? {
  var urlSessionTask: URLSessionTask?

  return try withTaskCancellationHandler {
    return try await withUnsafeThrowingContinuation { continuation in
      urlSessionTask = URLSession.shared.dataTask(with: url) { data, _, error in
        if let error = error {
          // Ideally translate NSURLErrorCancelled to CancellationError here
          continuation.resume(throwing: error)
        } else {
          continuation.resume(returning: data)
        }
      }
      urlSessionTask?.resume()
    }
  } onCancel: {
    urlSessionTask?.cancel() // runs immediately when cancelled
  }
}
```

DavidHorvath commented:
> Looks like the way to cancel an async/await function is https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md#cancellation-handlers
> 
> Here is the example in the like above.
> 
> ```
> func download(url: URL) async throws -> Data? {
>   var urlSessionTask: URLSessionTask?
> 
>   return try withTaskCancellationHandler {
>     return try await withUnsafeThrowingContinuation { continuation in
>       urlSessionTask = URLSession.shared.dataTask(with: url) { data, _, error in
>         if let error = error {
>           // Ideally translate NSURLErrorCancelled to CancellationError here
>           continuation.resume(throwing: error)
>         } else {
>           continuation.resume(returning: data)
>         }
>       }
>       urlSessionTask?.resume()
>     }
>   } onCancel: {
>     urlSessionTask?.cancel() // runs immediately when cancelled
>   }
> }
> ```

@4brunu Unfortunately this is not working the way it suppose to 😕 
I get a `Reference to captured var 'urlSessionTask' in concurrently-executing code` compile error.
https://forums.swift.org/t/how-to-cancel-a-publisher-when-using-withtaskcancellationhandler/49688

DavidHorvath commented:
@4brunu
I've added a new commit where I 
- Changed URLSessionDataTasks to URLSessionTasks
- Added support for request cancellation for alamofire library based on your comments
- Return the task in Result configuration

Still missing:
- Async/Await (I commented above that unfortunately that solution isn't working, I need to investigate further)

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -190,8 +199,9 @@ extension {{projectName}}API {
+
         return try await withCheckedThrowingContinuation { continuation in
-            {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
+            {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).executapiResponseQueue) { result in
```
@DavidHorvath I think this change was not intentional

DavidHorvath replied:
Oh... yeah it was unintentional

DavidHorvath replied:
Updated

4brunu commented on modules/openapi-generator/src/main/resources/swift5/libraries/urlsession/URLSessionImplementations.mustache:
```diff
@@ -170,11 +171,13 @@ private var credentialStore = SynchronizedDictionary<Int, URLCredential>()
 
             dataTask.resume()
-
+{{^useVapor}}{{^usePromiseKit}}{{^useResult}}{{^useAsyncAwait}}
```
@DavidHorvath I think this logic should be removed from here, and be handled in the `api.mustache`, as already is

DavidHorvath replied:
Yeah we don't need that, sorry, I left that in... I'll update in a minute

DavidHorvath replied:
Updated

4brunu commented on modules/openapi-generator/src/main/resources/swift5/libraries/urlsession/URLSessionImplementations.mustache:
```diff
@@ -170,11 +171,13 @@ private var credentialStore = SynchronizedDictionary<Int, URLCredential>()
             apiResponseQueue.async {
                 completion(.failure(ErrorResponse.error(415, nil, nil, error)))
-            }
+            }{{^useVapor}}{{^usePromiseKit}}{{^useResult}}{{^useAsyncAwait}}
```
@DavidHorvath I think this logic should be removed from here, and be handled in the api.mustache, as already is

DavidHorvath replied:
Yeah we don't need that, sorry, I left that in... I'll update in a minute

4brunu replied:
No problem 🙂 Thanks for your help

DavidHorvath replied:
Updated

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -157,8 +161,9 @@ extension {{projectName}}API {
+        var task: URLSessionTask?
         return Future<{{{returnType}}}{{^returnType}}Void{{/returnType}}, Error> { promise in
-            {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
+            {{#useURLSession}}dataTask = {{/useURLSession}}{{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
```
@DavidHorvath I think here we can remove the checks on the URLSession, because now is alamofire also supports cancellation

DavidHorvath replied:
fixed

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -172,7 +177,11 @@ extension {{projectName}}API {
                 }
             }
-        }.eraseToAnyPublisher()
+        }{{#useURLSession}}
```
@DavidHorvath I think here we can remove the checks on the URLSession, because now is alamofire also supports cancellation

DavidHorvath replied:
fixed

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -122,8 +123,8 @@ extension {{projectName}}API {
     {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} class func {{operationId}}({{#allParams}}{{paramName}}: {{#isEnum}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{^isContainer}}{{{datatypeWithEnum}}}_{{operationId}}{{/isContainer}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{/isEnum}}{{^required}}? = nil{{/required}}{{^-last}}, {{/-last}}{{/allParams}}{{#hasParams}}, ...
-        return Observable.create { observer -> Disposable in
-            {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
+        {{#useURLSession}}return {{/useURLSession}}Observable.create { observer -> Disposable in
```
@DavidHorvath I think here we can remove the checks on the URLSession, because now is alamofire also supports cancellation

DavidHorvath replied:
fixed

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -138,7 +139,10 @@ extension {{projectName}}API {
             }
-            return Disposables.create()
+            
+            return Disposables.create{{^useURLSession}}(){{/useURLSession}}{{#useURLSession}} {
```
@DavidHorvath I think here we can remove the checks on the URLSession, because now is alamofire also supports cancellation

DavidHorvath replied:
Oh I was neglectful... Sorry about this and thanks for your comments!

DavidHorvath replied:
fixed

4brunu commented:
@DavidHorvath I think I got `async/await` working with cancellation
```
    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
    open class func fakeOuterBooleanSerialize1(body: Bool? = nil, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) async throws -> Bool {
        var task: URLSessionTask?
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                task = fakeOuterBooleanSerializeWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
                    switch result {
                    case let .success(response):
                        continuation.resume(returning: response.body!)
                    case let .failure(error):
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { [task] in
            task?.cancel()
        }
    }
```

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -190,6 +199,7 @@ extension {{projectName}}API {
     @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
     {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} class func {{operationId}}({{#allParams}}{{paramName}}: {{#isEnum}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{^isContainer}}{{{datatypeWithEnum}}}_{{operationId}}{{/isContainer}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{/isEnum}}{{^required}}? = nil{{/required}}{{^-last}}, {{/-last}}{{/allParams}}{{#hasParams}}, ...
+
         return try await withCheckedThrowingContinuation { continuation in
```
@DavidHorvath  Can you please try this?
```
    {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} class func {{operationId}}({{#allParams}}{{paramName}}: {{#isEnum}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{^isContainer}}{{{datatypeWithEnum}}}_{{operationId}}{{/isContainer}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{/isEnum}}{{^required}}? = nil{{/required}}{{^-last}}, {{/-last}}{{/allParams}}{{#hasParams}}, {{/hasParams}}apiResponseQueue: DispatchQueue = {{projectName}}API.apiResponseQueue) async throws{{#returnType}} -> {{{returnType}}}{{/returnType}} {
        var task: URLSessionTask?
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                task = {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
                    switch result {
                    {{#returnType}}
                    case let .success(response):
                        continuation.resume(returning: response.body!)
                    {{/returnType}}
                    {{^returnType}}
                    case .success:
                        continuation.resume(returning: ())
                    {{/returnType}}
                    case let .failure(error):
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { [task] in
            task?.cancel()
        }
    }
```

4brunu replied:
I hope I got it right, because I didn't tested it, but I tested the swift code that I send I minute ago

DavidHorvath replied:
It wasn't very straightforward at first how to do it but finally I was able to test our generated async await APIClient.
I've displayed the response on the UI or a `Cancelled` text if the task had been canceled.

``` swift
    func load() async throws -> String {
        let fetchTask = Task { () -> String in
            do {
                let response = try await TestAPI.asyncTest()
                if Task.isCancelled { return "Cancelled" }

                return response
            } catch {
                return "Error"
            }
        }

        fetchTask.cancel()

        return await fetchTask.value
    }

    open class func asyncTest(apiResponseQueue: DispatchQueue = TestAPI.apiResponseQueue) async throws -> String {
        var task: URLSessionTask?
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                task = testWithRequestBuilder().execute(apiResponseQueue) { result in
                    switch result {
                    case let .success(response):
                        continuation.resume(returning: response.body!)
                    case let .failure(error):
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { [task] in
            print("Task Cancelled")
            task?.cancel()
        }
    }

```

And there was a `Task Cancelled` console message and a `Cancelled` text on the UI as well, so it works! 🙂 🚀

4brunu replied:
Nice, good to know 👍

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -157,8 +161,9 @@ extension {{projectName}}API {
+        var task: URLSessionTask?
         return Future<{{{returnType}}}{{^returnType}}Void{{/returnType}}, Error> { promise in
-            {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
+            dataTask = {{operationId}}WithRequestBuilder({{#allParams}}{{paramName}}: {{paramName}}{{^-last}}, {{/-last}}{{/allParams}}).execute(apiResponseQueue) { result in
```
@DavidHorvath this `dataTask` doesn't match with `task` variable bellow, could you please rename this to `task`?

4brunu commented on modules/openapi-generator/src/main/resources/swift5/libraries/alamofire/AlamofireImplementations.mustache:
```diff
@@ -78,7 +78,8 @@ private var managerStore = SynchronizedDictionary<String, Alamofire.Session>()
 
-    override {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} func execute(_ apiResponseQueue: DispatchQueue = {{projectName}}API.apiResponseQueue, _ completion: @escaping (_ result: Swift.Result<Response<T>, ErrorResponse>) -> Void) {
+    @discardableResult
+    override {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}open{{/nonPublicApi}} func execute(_ apiResponseQueue: DispatchQueue = {{projectName}}API.apiResponseQueue, _ completion: @escaping (_ result: Swift.Result<Response<T>, ErrorResponse>) -> Void) -> URLSessionTask {
```
@DavidHorvath here the return type should be `URLSessionTask?`

DavidHorvath replied:
Fixed

4brunu commented on modules/openapi-generator/src/main/resources/swift5/api.mustache:
```diff
@@ -220,8 +230,9 @@ extension {{projectName}}API {
     @available(*, deprecated, message: "This operation is deprecated.")
     {{/isDeprecated}}
+    @discardableResult
     open class func {{operationId}}({{#allParams}}{{paramName}}: {{#isEnum}}{{#isContainer}}{{{dataType}}}{{/isContainer}}{{^isContainer}}{{{datatypeWithEnum}}}_{{operationId}}{{/isContainer}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{/isEnum}}{{^required}}? = nil{{/required}}{{^-last}}, {{/-last}}{{/allParams}}{{#hasParams}}, {{/hasParams}}apiResponseQueue: DispatchQueue = {{projectName}}API.apiRespo ...
```
@DavidHorvath here we are missing the return type `URLSessionTask?`

4brunu replied:
@DavidHorvath can you please also fix this one please?

4brunu commented:
@DavidHorvath I don't want to be annoying, but I got one last improvement that it would be nice, but if you prefer I can do it in a separated PR.
In the async/await we could check for cancellation, like so:
```
    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
    open class func call123testSpecialTags(body: Client, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) async throws -> Client {
        var task: URLSessionTask?
        return try await withTaskCancellationHandler {
            try Task.checkCancellation() // HERE
            return try await withCheckedThrowingContinuation { continuation in // also add return in the beginning of the line
                guard !Task.isCancelled else { // and HERE
                  continuation.resume(throwing: CancellationError())
                  return
                }
                task = call123testSpecialTagsWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
                    switch result {
                    case let .success(response):
                        continuation.resume(returning: response.body!)
                    case let .failure(error):
                        continuation.resume(throwing: error)
                    }
                }
            }
        } onCancel: { [task] in
            task?.cancel()
        }
    }
```
What do you think?
I got this idea from here
https://www.pointfree.co/episodes/ep154-async-refreshable-composable-architecture (in the first exercise that use withTaskCancellationHandler)

DavidHorvath commented:
> @DavidHorvath I don't want to be annoying, but I got one last improvement that it would be nice, but if you prefer I can do it in a separated PR. In the async/await we could check for cancellation, like so:
> 
> ```
>     @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
>     open class func call123testSpecialTags(body: Client, apiResponseQueue: DispatchQueue = PetstoreClientAPI.apiResponseQueue) async throws -> Client {
>         var task: URLSessionTask?
>         return try await withTaskCancellationHandler {
>             try Task.checkCancellation() // HERE
>             return try await withCheckedThrowingContinuation { continuation in // also add return in the beginning of the line
>                 guard !Task.isCancelled else { // and HERE
>                   continuation.resume(throwing: CancellationError())
>                   return
>                 }
>                 task = call123testSpecialTagsWithRequestBuilder(body: body).execute(apiResponseQueue) { result in
>                     switch result {
>                     case let .success(response):
>                         continuation.resume(returning: response.body!)
>                     case let .failure(error):
>                         continuation.resume(throwing: error)
>                     }
>                 }
>             }
>         } onCancel: { [task] in
>             task?.cancel()
>         }
>     }
> ```
> 
> What do you think? I got this idea from here https://www.pointfree.co/episodes/ep154-async-refreshable-composable-architecture (in the first exercise that use withTaskCancellationHandler)

@4brunu Sure I can add that as well

4brunu reviewed (approved):
Looks good to me 👍 
@DavidHorvath Thank you very much for the two awesome PRs, that were very impactful in the Swift Generator, and for your quality PRs.
Thanks you also for your patience and for doing the required changes to merge the PRs.
Are you planning to do open more PRs? 😃 
Feel free to contribute any time you want 👍
